cis194/week2/stevemao/LogAnalysis.hs

```diff
@@ -22,3 +22,27 @@ parseMessage a = case words a of
 
 parse :: String -> [LogMessage]
 parse = fmap parseMessage . lines
+
+insert :: LogMessage -> MessageTree -> MessageTree
+insert (Unknown _) tree = tree
+insert a Leaf = Node Leaf a Leaf
+insert log1@(LogMessage _ ts1 _) (Node left log2@(LogMessage _ ts2 _) right)
+  | ts1 > ts2 = Node left log2 (insert log1 right)
+  | otherwise = Node (insert log1 left) log2 right
+
+build :: [LogMessage] -> MessageTree
+build = foldr (\logMsg tree -> insert logMsg tree) Leaf
+
+inOrder :: MessageTree -> [LogMessage]
+inOrder Leaf = []
+inOrder (Node left a right) = inOrder left ++ [a] ++ inOrder right
+
+filterFn :: LogMessage -> Bool
+filterFn (LogMessage (Error severity) _ _) = severity > 50
+filterFn _ = False
+
+mapper :: LogMessage -> String
+mapper (LogMessage _ _ message) = message
```

Comment:
We can name this getMessage.
Bit more importantly this function is partial!

Reply:
It shouldn't be possible to match `Unknown`. Let me have a look how this should be done.